### PR TITLE
feat(backend): human-readable file ids

### LINKED ISSUE #6907: feat(backend): `fileIds` should actually be accessions or at least human-friendly and not UUIDs as they are exposed to users
@corneliusroemer thoughts from a #6883 comment:

> @anna-parker suggested in a conversation that we use a more human-friendly and concise accession-like id for files instead of UUIDs. I think we chose UUIDs because it was easy to do, not expecting to surface them to users in any way.
> It shouldn't be too hard to issue more human friendly ids by the backend by adding a new table to the db and having a service in the backend provide those friendlier ids to the service that currently just picks UUIDs.
> Ids could be anything that's unique, but we could make them something like FILE_AB1X similar to how we issue sequence accessions but probably without fixed length and without checksum character.

### PR DESCRIPTION
**Pull request by maverbiest**
resolves #6907 

Users see `fileIds` in the `files.<category>` column of metadata files produced when they download original metadata for a revision. If they want to edit the `files` column, they need to visually parse the space-separated list of `file:fileId` where `fileId` is a 36 character UUID, and modify it where appropriate. This PR makes it so file ids have the format `FILE_<accession>`, e.g., `FILE_00G3`. This is similar to how sequence accessions are made, except they're shorter and don't have a checksum character appended (see linked issue).

_**Note:** existing deployments that already have file sharing enabled don't have to change existing file ids. The uuid-style ids are still supported, they are now just treated as strings (see 'Manual testing' below)_

### Summary

- `FileId` changes from a `UUID` to a `String`
- Added a sequence to the DB to track file ids
- Extract base 34 encoding functionality into `Base34.kt`
- Added `createFileEntries` to `FilesDatabaseService.kt` to do batch creation of files. Let me know if this is beyond the scope, but previous state was to have a loop where each file insert was a separate DB transaction (although most usage will prob go through multipart uploads, which can't be batched since it calls out to S3...).

### Why does this touch so many files?

- Refactoring `FileId` from `UUID` to `String` meant all call sites that specified `UUID` or did `UUID.fromString()` etc. had to be updated
- Docs/swagger showed example file ids that were 36-char UUIDs, I updated these to the new format for consistency

#### First test

I did a submission with raw reads on the preview, downloaded original metadata, and confirmed that the file id is now in the new format:

grafik

#### Second test

To test that legacy uuid-style IDs will keep working, I uploaded two files, changed the id for one of them to uuid-style, and submitted sequences pointing to those raw reads (see 'Testing details' below). When I download original metadata for these two sequences, the uuid-style and new id show up next to each other without issue:

grafik

I also confirmed that I can still download the renamed raw reads file from the website.

<details>
<summary>Testing details</summary>
All file upload requests on the current branch's preview will get the new FILE_xxxx style ids. To test that the system will keep supporting legacy uuid-style ids, I had to edit the id of an uploaded raw reads file after submission. I did this through the API as follows:

**Set backend url, get token**
```sh
BACKEND=https://backend-readable-file-ids.loculus.org
TOKEN=$(curl -X POST "https://authentication-readable-file-ids.loculus.org/realms/loculus/protocol/openid-connect/token" -d 'username=testuser&password=testuser&grant_type=password&client_id=backend-client' | jq -r .access_token)
```

**Request upload of two files**
(I made a test group for testuser with groupId=2 on the preview first)
```sh
curl -s -X POST "$BACKEND/files/request-upload?groupId=2&numberFiles=2" -H "Authorization: Bearer $TOKEN"
```

**PUT two raw reads files to the returned presigned urls**
(Saved presigned urls into URL1 and URL2 first)
```sh
curl -X PUT -H 'If-None-Match: *' -T raw_reads_1.fq.gz "$URL1"
curl -X PUT -H 'If-None-Match: *' -T raw_reads_2.fq.gz "$URL2"
```

**Rename one of the files on S3 and in the DB**
```sh
kubectl exec -n prev-readable-file-ids deploy/minio -- sh -c 'mc alias set l http://localhost:9000 "$MINIO_ROOT_USER" "$MINIO_ROOT_PASSWORD" && mc mv l/loculus-preview-private/files/FILE_0TJS l/loculus-preview-private/files/8d8ac610-566d-4ef0-9c22-186b2a5ed793'
Added `l` successfully.
`l/loculus-preview-private/files/FILE_0TJS` -> `l/loculus-preview-private/files/8d8ac610-566d-4ef0-9c22-186b2a5ed793`
┌────────────┬─────────────┬──────────┬────────────┐
│ Total      │ Transferred │ Duration │ Speed      │
│ 254.88 KiB │ 254.88 KiB  │ 00m00s   │ 9.58 MiB/s │
└────────────┴─────────────┴──────────┴────────────┘

kubectl exec -n prev-readable-file-ids deploy/loculus-database -- psql -U postgres -d loculus -c "UPDATE files SET id = '8d8ac610-566d-4ef0-9c22-186b2a5ed793' WHERE id = 'FILE_0TJS';"
UPDATE 1
```

**Submit files**
(I had prepared some test metadata and sequences for west-nile)
```sh
curl -X POST "$BACKEND/west-nile/submit?groupId=2&dataUseTermsType=OPEN" -H "Authorization: Bearer $TOKEN" -F metadataFile=@west_nile_metadata.tsv -F sequenceFile=@west_nile_sequences.fasta
[{"accession":"LOC_001JZ7H","version":1,"submissionId":"test_WNV/mosquito/Larimer-CO-USA/2016-06-27/W1819"},{"accession":"LOC_001JZ8F","version":1,"submissionId":"test_WNV/mosquito/Polk-IA-USA/2017-07-17/W1959"}]%
```

Then, I went to the preview website, went to west-nile -> review sequences -> approved the sequences. They showed up on the preview a little while later, and then I downloaded original metadata to confirm that both styles of file id still work

</details>

#### Third test

Finally, I ran additional tests that @anna-parker suggested:
1. I created a new branch off of main: https://github.com/loculus-project/loculus/pull/7306
2. uploaded files there, confirmed they got UUID identifiers
3. cherry-picked the changes from this PR onto that branch 
4. uploaded more files.
5. revised a submission from the first round by adding an additional raw reads file

The result was that the first files I submitted got UUID ids, then the new commit with the SQL migration changed these to strings, and the second round of files I uploaded got the new ids:

grafik

### PR Checklist
- [x] All necessary documentation has been adapted.
- [x] The implemented feature is covered by appropriate, automated tests.
- [x] Any manual testing that has been done is documented (i.e. what exactly was tested?)

🚀 Preview: Add `preview` label to enable

### CODE REVIEW

File: backend/src/test/kotlin/org/loculus/backend/utils/MetadataEntryTest.kt [hunk 1]
```diff
@@ -281,31 +279,6 @@ class MetadataEntryTest {
         assertThat(exception.message, containsString("missing a file name"))
     }
 
-    @Test
```
Comment (maverbiest) [inline backend/src/test/kotlin/org/loculus/backend/utils/MetadataEntryTest.kt:284]: Deleted as test file ids are no longer UUIDs

File: backend/src/main/kotlin/org/loculus/backend/controller/FilesController.kt [hunk 2]
```diff
@@ -173,19 +170,19 @@ class FilesController(
...
         filesPreconditionValidator.validateUserIsAllowedToUploadFileForGroup(groupId, authenticatedUser)
-        val response = mutableListOf<FileIdAndMultipartWriteUrl>()
-        repeat(numberFiles) {
-            val fileId = generateFileId()
+        if (numberFiles < 1) {
+            throw BadRequestException("Number of files must be at least 1")
+        }
+        return generateFileIds(numberFiles).map { fileId ->
             val multipartUploadHandler = s3Service.initiateMultipartUploadAndCreateUrlsToUpload(fileId, numberParts)
             filesDatabaseService.createFileEntry(
```
Comment (maverbiest) [inline backend/src/main/kotlin/org/loculus/backend/controller/FilesController.kt:176]: This still uses `createFileEntry` since `s3Service.initiateMultipartUploadAndCreateUrlsToUpload` can't be batched

File: backend/src/main/kotlin/org/loculus/backend/service/GenerateAccessionFromNumberService.kt [hunk 3]
```diff
@@ -2,25 +2,16 @@ package org.loculus.backend.service
...
+import org.loculus.backend.utils.CODE_POINTS
+import org.loculus.backend.utils.base34Encode
 import org.springframework.beans.factory.annotation.Autowired
 import org.springframework.stereotype.Service
 
 @Service
 class GenerateAccessionFromNumberService(@Autowired val backendConfig: BackendConfig) {
 
     fun generateCustomId(sequenceNumber: Long): String {
-        val base34Digits: MutableList<Char> = mutableListOf()
```
Comment (maverbiest) [inline backend/src/main/kotlin/org/loculus/backend/service/GenerateAccessionFromNumberService.kt:12]: Moved to `Base34.kt`

**Comment by maverbiest** [comment]
@claude review

File: backend/src/main/kotlin/org/loculus/backend/utils/GenerateFileId.kt [hunk 4]
```diff
@@ -1,5 +1,9 @@
 package org.loculus.backend.utils
 
-import java.util.UUID
+import org.loculus.backend.service.files.FileId
 
-fun generateFileId() = UUID.randomUUID()
+private const val FILE_ID_PREFIX = "FILE_"
+private const val FILE_ID_MIN_SERIAL_LENGTH = 4
+
+fun generateFileIds(count: Int): List<FileId> = getNextSequenceNumbers("file_id_sequence", count)
```
[the remark this answers is not in the document]
Comment (maverbiest) [reply backend/src/main/kotlin/org/loculus/backend/utils/GenerateFileId.kt:8]: I don't think this is needed, we don't really expect users to hand-type these, right? They're more likely to either remove them fully or leave them alone

File: backend/src/main/kotlin/org/loculus/backend/controller/FilesController.kt [hunk 5]
```diff
@@ -173,19 +170,19 @@ class FilesController(
         numberParts: Int = 1,
     ): List<FileIdAndMultipartWriteUrl> {
         filesPreconditionValidator.validateUserIsAllowedToUploadFileForGroup(groupId, authenticatedUser)
-        val response = mutableListOf<FileIdAndMultipartWriteUrl>()
-        repeat(numberFiles) {
-            val fileId = generateFileId()
+        if (numberFiles < 1) {
+            throw BadRequestException("Number of files must be at least 1")
+        }
```
[the remark this answers is not in the document]
Comment (maverbiest) [reply backend/src/main/kotlin/org/loculus/backend/controller/FilesController.kt]: Added the check to `filesPreconditionValidator`, rather than adding Min and Max to the controller so we have nicer errors.

I didn't include a cap on how many files can be created per request, but would be happy to add if people feel it's needed

**Comment by theosanderson** [comment]
@codex review

File: backend/src/main/kotlin/org/loculus/backend/utils/MetadataEntry.kt [hunk 6]
```diff
@@ -162,23 +160,11 @@ private fun extractAndValidateFileIdAndName(
...
-    } catch (e: IllegalArgumentException) {
+    if (fileId.isEmpty()) {
         throw UnprocessableEntityException(
             "In metadata file: record #$recordNumber with id '$submissionId': " +
-                "file entry '$token' in column '$header' has an invalid file ID '$fileIdString'. " +
-                "Expected a UUID.",
-        )
-    }
-
-    if (fileIdString.length != FILE_ID_LENGTH) {
```
Comment (tombch) [inline backend/src/main/kotlin/org/loculus/backend/utils/MetadataEntry.kt:186]: The check here for UUID length was added because we found it very easy to accidentally delete one character of the UUID when revising a metadata file and got stumped for a few minutes with the `file ID not found` message. (should have commented this, sorry!)

Maybe that check is still useful (although a bit hacky?). Although I guess it would be more noticeable now for a user if they accidentally delete a character for this shorter ID?

Comment (tombch) [reply backend/src/main/kotlin/org/loculus/backend/utils/MetadataEntry.kt:186]: I think this would be fine without a length check now the file IDs aren't UUIDs, if it became a pain point for people when revising metadata maybe something to reconsider

Comment (maverbiest) [reply backend/src/main/kotlin/org/loculus/backend/utils/MetadataEntry.kt:186]: Yeah makes sense! I'm also hoping that the shorter file ids will alleviate this. If other feel it would be better I could still include a check on the minimal id length (they can now grow in size if we have enough files)

Comment (anna-parker) [reply backend/src/main/kotlin/org/loculus/backend/utils/MetadataEntry.kt:186]: We dont cap the size of the fileIds and as the Ids we generate can have gaps we might get over 4 characters in length - so yes we should remove the string length check :-)

**Comment by anna-parker** [comment]
Am I correct in understanding that loculus users that already have files with UUIDs will not need to do an additional migration as the UUIDs will just become strings - and for those users the old file IDs will be UUIDs and new fileIds will have the format <FILE_xxxx>?

**Comment by maverbiest** [comment]
> Am I correct in understanding that loculus users that already have files with UUIDs will not need to do an additional migration as the UUIDs will just become strings

@anna-parker yes exactly, deployments that had additional files enabled will have UUIDs and FILE_xxxx in the same DB column going forward, but now they're just interpreted as strings.

EDIT: I will come up with a way to test this on the preview

**Comment by anna-parker** [comment]
@maverbiest yeah I think it would be great to test this out and add an explanation to the PR description

**Comment by maverbiest** [comment]
@anna-parker I confirmed on the preview that uuid-style ids will keep working! I put a note in the PR description saying old file ids do not need to be updated, and also added a description of how I tested this. Hope it's clear enough!

**Comment by anna-parker** [comment]
@maverbiest thanks for testing and explaining the steps - I think this is probably ok but I would have probably tested by:

1. creating a branch off main, setting the db to persistent
2. let sequences get ingested with annotation files and then submit a couple more files to the preview
3. commit all these changes to that branch
4. confirm the db migration works and there is nothing else we potentially missed, i.e. can revise a sequence that has a file attached to it etc. 

This way we would cover the migration and ensure there isnt anything in the code we missed - I know its more work but it might be a good thing to still do if you have time.

File: backend/src/test/kotlin/org/loculus/backend/utils/MetadataEntryTest.kt [hunk 7]
```diff
@@ -204,8 +202,8 @@ class MetadataEntryTest {
             equalTo(
                 mapOf(
                     "raw_reads" to listOf(
-                        FileIdAndName(UUID.fromString(fileId1), "reads_1.fq"),
-                        FileIdAndName(UUID.fromString(fileId2), "reads_2.fq"),
+                        FileIdAndName(fileId1, "reads_1.fq"),
```
Comment (anna-parker) [inline backend/src/test/kotlin/org/loculus/backend/utils/MetadataEntryTest.kt:206]: the fileIds used in the tests here are actually in the old UUID format - Im not sure if this is intentional, maybe it would be good to have both formats

**Comment by maverbiest** [comment]
> but I would have probably tested by:

@anna-parker yeah that would be a more realistic testing procedure. I'll get back to you once I've done that!

File: backend/src/test/kotlin/org/loculus/backend/utils/MetadataEntryTest.kt (hunk 7 above)
Comment (maverbiest) [reply backend/src/test/kotlin/org/loculus/backend/utils/MetadataEntryTest.kt:206]: Yes I was debating this a bit, I thought it might be good to have some old file ids in the tests but I don't know if it's too confusing to have the two formats when someone looks at this in the future. 

I should probably have all tests use the new `dummyFileId()` that returns the new style, and add a dedicated test that covers the legacy format

**Review by anna-parker — APPROVED** [review]
this looks great - if no issues come up while testing feel free to merge, you can tag me if you need to make more changes and Ill take a look :-)

**Comment by maverbiest** [comment]
@anna-parker I ran the tests you suggested and all seems good! (see updated PR description)

I also decided to clean up the tests so none of them use UUID strings anymore, except for a new test in `SubmitEndpointFileSharingTest.kt` where I explicitly test for UUID and FILE_xxxx ids existing next to each other. Would you like to take a look?

**Comment by anna-parker** [comment]
@maverbiest amazing thanks! this looks good to merge!
